persistence/data_models_json_fields.go

```diff
@@ -91,6 +91,164 @@ func (s *StateExecutionSequenceMapsJson) CompleteNewStateExecution(stateId strin
+	StateToCommandsMap map[string][]xdbapi.LocalQueueCommand `json:"stateToCommandsMap"`
+	// { queue_name_1: [dedupId_1, dedupId_2, ...], queue_name_2: [dedup_id, ...], ... }
+	UnconsumedMessageQueueDedupIdsMap map[string][]uuid.UUID `json:"unconsumedMap"`
+}
+
+func NewStateExecutionWaitingQueues() StateExecutionWaitingQueuesJson {
+	return StateExecutionWaitingQueuesJson{
+		StateToCommandsMap:                map[string][]xdbapi.LocalQueueCommand{},
+		UnconsumedMessageQueueDedupIdsMap: map[string][]uuid.UUID{},
+	}
+}
+
+func (s *StateExecutionWaitingQueuesJson) ToBytes() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+func NewStateExecutionWaitingQueuesFromBytes(bytes []byte) (StateExecutionWaitingQueuesJson, error) {
+	var waitingQueuesJson StateExecutionWaitingQueuesJson
+	err := json.Unmarshal(bytes, &waitingQueuesJson)
+	return waitingQueuesJson, err
+}
+
+func (s *StateExecutionWaitingQueuesJson) AddNewLocalQueueCommandForStateExecution(
+	stateExecutionId StateExecutionId, command xdbapi.LocalQueueCommand) {
+	if command.GetCount() == 0 {
+		command.Count = xdbapi.PtrInt32(1)
+	}
+
+	stateExecutionIdKey := stateExecutionId.GetStateExecutionId()
+
+	s.StateToCommandsMap[stateExecutionIdKey] = append(s.StateToCommandsMap[stateExecutionIdKey], command)
+}
+
+// Consume return (StateExecutionId string, dedupIds) where the StateExecutionId consumes messages of these dudupIds.
+//
+// E.g., given StateToCommandsMap as:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+//	state_3, 1: (q1: 1), (q2, 2)
+//
+// If receiving the queue `q1`, then state_3, 1 will consume the queue `q1`, and StateToCommandsMap becomes:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+//	state_3, 1: (q2, 2)
+//
+// If receiving the queue `q2` for twice, then state_3, 1 will consume the two queue `q2`, and StateToCommandsMap becomes:
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+func (s *StateExecutionWaitingQueuesJson) Consume(message LocalQueueMessageInfoJson) (*string, []uuid.UUID) {
+	s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] = append(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName], message.DedupId)
+
+	for stateExecutionIdKey, commands := range s.StateToCommandsMap {
+		for i, command := range commands {
+			if command.GetQueueName() != message.QueueName || int(command.GetCount()) > len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) {
+				continue
+			}
+
+			consumedDedupIds := s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][:int(command.GetCount())]
+
+			s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] =
+				s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][int(command.GetCount()):]
+			if len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) == 0 {
+				delete(s.UnconsumedMessageQueueDedupIdsMap, message.QueueName)
+			}
+
+			s.StateToCommandsMap[stateExecutionIdKey] = append(s.StateToCommandsMap[stateExecutionIdKey][:i],
```

Comment:
Wow, this is so concise. I really like it after understanding it. 
Though it took me a couple minutes to understand:
* it's different from the regular pattern: `slice = append(slice, new eles)`, but just look so closed!!! LOL
* it's not very usual to just update the map/slice when iterating through it(similar in Java, you usually don't update the array when iterating it)

So I would recommend:
* Move this update outside of the for loop. Since it's the last thing to do before return. We could just break here (maybe set a "foundMatched" bool, and do it outside.
* Add a friendly comment/note about this cool operation!

Reply:
Adding a comment sounds good to me. 

If moving the logic outside, then we need to record 2 break points, one for command, one for state, I feel it's a little more cumbersome to write.

Reply:
You can label the first for loop like this https://stackoverflow.com/questions/51996175/how-to-break-out-of-nested-loops-in-go